<launch>
  <!-- 0. JAKA 驱动与 URDF TF -->
  <arg name="ip" default="192.168.1.100" />
  <arg name="urdf_file" default="$(find jaka_close_contro)/urdf/jaka_zu3.urdf" />

  <include file="$(find jaka_close_contro)/launch/jaka_sdk_bringup.launch">
    <arg name="ip" value="$(arg ip)" />
    <arg name="urdf_file" value="$(arg urdf_file)" />
  </include>

  <!-- 1. ZED2i 相机 -->
  <include file="$(find zed_wrapper)/launch/zed2i.launch">
    <arg name="svo_file" value="" />
  </include>

  <!-- 2. ArUco 检测 + 分流 + 世界板 + 多面融合，与标定流程一致 -->
  <node pkg="jaka_close_contro"
        type="cube_aruco_detector_node"
        name="cube_aruco_detector"
        output="screen">
    <param name="image_topic"       value="/zed2i/zed_node/left/image_rect_color"/>
    <param name="camera_info_topic" value="/zed2i/zed_node/left/camera_info"/>
    <param name="fiducial_topic"    value="/fiducial_transforms"/>
    <param name="default_marker_size" value="0.04"/>
    <rosparam param="marker_sizes">
      "0": 0.08  <!-- World Tag (80mm) -->
      "10": 0.04 <!-- Cube face 1 (40mm) -->
      "11": 0.04 <!-- Cube face 2 -->
      "12": 0.04 <!-- Cube face 3 -->
      "13": 0.04 <!-- Cube face 4 -->
    </rosparam>
  </node>

  <node pkg="jaka_close_contro"
        type="fiducial_relay_node"
        name="fiducial_relay"
        output="screen">
    <param name="world_id"   value="0"/>
    <rosparam param="tool_ids">["10", "11", "12", "13"]</rosparam>
    <param name="input_topic" value="/fiducial_transforms"/>
    <param name="world_topic" value="/world_fiducials"/>
    <param name="tool_topic"  value="/tool_fiducials"/>
  </node>

  <node pkg="jaka_close_contro"
        type="world_tag_node"
        name="world_tag_node"
        output="screen">
    <param name="world_fiducial_id" value="0"/>
    <param name="fiducial_topic"    value="/world_fiducials"/>
    <param name="world_frame"       value="world"/>
    <param name="camera_frame"      value="zed2i_left_camera_optical_frame"/>
    <param name="alpha_pos"         value="0.3"/>
    <param name="alpha_rot"         value="0.3"/>
  </node>

  <node pkg="jaka_close_contro"
        type="cube_multi_face_fusion_node"
        name="cube_multi_face_fusion"
        output="screen">
    <param name="faces_yaml"             value="$(find jaka_close_contro)/config/cube_faces_current.yaml"/>
    <param name="fiducial_topic"         value="/tool_fiducials"/>
    <param name="cube_frame"             value="cube_center"/>
    <param name="camera_frame_default"   value="zed2i_left_camera_optical_frame"/>
    <param name="alpha_pos"              value="0.3"/>
    <param name="alpha_rot"              value="0.3"/>
    <param name="max_step_pos"           value="0.03"/>
    <param name="max_step_deg"           value="15.0"/>
    <param name="timeout_sec"            value="0.5"/>
    <param name="publish_tool_tf"        value="false"/>
  </node>

  <!-- 3. 世界-机器人标定节点（提供采样/求解服务） -->
  <node pkg="jaka_close_contro"
        type="world_robot_calibration_node"
        name="world_robot_calibration_node"
        output="screen">
    <param name="world_frame"       value="world" />
    <param name="robot_base_frame"  value="Link_0" />
    <param name="tool_frame"        value="Link_6" />
    <param name="camera_frame"      value="zed2i_left_camera_optical_frame" />
    <param name="cube_frame"        value="cube_center" />
    <param name="cube_offset_z_m"   value="0.077" />
    <param name="min_samples"       value="16" />
    <param name="publish_tf"        value="false" />
  </node>

  <!-- 4. 自动采样/求解管理节点 -->
  <node pkg="jaka_close_contro"
        type="world_robot_autocalib_manager_node"
        name="world_robot_autocalib_manager"
        output="screen">
    <param name="min_translation" value="0.05" />
    <param name="min_angle_deg"   value="15.0" />
    <param name="min_samples"     value="16" />
    <param name="auto_solve"      value="true" />
    <param name="extrinsic_yaml"  value="$(find jaka_close_contro)/config/world_robot_extrinsic.yaml" />
    <param name="cube_topic"      value="/cube_center_fused" />
    <param name="world_frame"     value="world" />
    <param name="camera_frame"    value="zed2i_left_camera_optical_frame" />
    <param name="cube_frame"      value="cube_center" />
    <param name="robot_base_frame" value="Link_0" />
  </node>

  <!-- 5. 标定轨迹自动执行（速度限制 15%） -->
  <node pkg="jaka_close_contro"
        type="world_robot_calib_motion_node"
        name="world_robot_calib_motion"
        output="screen">
    <param name="arm_name"         value="jaka1" />
    <param name="calib_pose_csv"   value="$(find jaka_close_contro)/config/jaka1_world_robot_calibration_pose.csv" />
    <param name="speed_scale" value="0.15" />
    <param name="linear_speed_mm_s" value="80.0" />
    <param name="linear_acc_mm_s2" value="200.0" />
    <param name="pre_sample_wait_sec" value="5.0" />
    <param name="post_sample_wait_sec" value="2.0" />
  </node>
</launch>
